### ISSUE TITLE
Created a PyTorch implementation for the encoder layers

### ISSUE THREAD

**Issue by robin-poelmans** [jrwnter/cddd#33]
Hi,

I'm a little late to the party, but great work on the CDDD paper! 
I'm planning to use the CDDD pre-trained encoder layers in my own AI project, but here I stumbled on the issue that CDDD is written in TensorFlow 1.x, while the rest of my project will be built in PyTorch.

So, I decided to make my own PyTorch implementation of CDDD, where I recreated the network from scratch and then manually set the weights that I took from the TensorFlow model. This was actually surprisingly difficult, since PyTorch uses a different implementation for the GRU cell, where the order of the Hadamard product is changed when calculating the new candidate vector. My implementation now reproduces the results from the TF 1.x model with very high accuracy.
Since it took me a few days to create a new custom GRU function for PyTorch mimicking the behavior of TF 1.x and to rebuild the entire CDDD encoder, I wanted to share my work with the community:

[](url)https://github.com/robin-poelmans/CDDD_torch/tree/main

**Example command**:
python cddd_encoder_torch.py -i test2.smi -o test.csv --weights_dir '/media/drives/drive1/robin/cddd/default_model/weights' --batch_size 32

Input = one SMILES string per row
Output = csv starting with SMILES, then all 512 features

To create the weights_dir, first execute the following code in the original CDDD environment in the cddd/default_model directory:

```# In original TensorFlow environment:
import tensorflow as tf
import numpy as np
import os

# Load the model
sess = tf.Session()
saver = tf.train.import_meta_graph('model.ckpt.meta')
saver.restore(sess, 'model.ckpt')
graph = sess.graph

# Get all variables
variables = tf.trainable_variables()

# Save each variable as a separate NumPy file
os.mkdir('weights')
for var in variables:
    value = sess.run(var)
    np.save(f'weights/{var.name.replace("/", "_").replace(":", "_")}.npy', value)
```

I'm sure there's still room for improvement, but I hope other people will be able to use this implementation as well. I have currently only tested it on cpu, so I'm not sure yet how it will run on gpu
